x-pack/plugins/security_solution/public/common/components/navigation/breadcrumbs/index.ts

```diff
@@ -64,156 +71,107 @@ export const useSetBreadcrumbs = () => {
-  spyState != null && spyState.pageName === SecurityPageName.network;
-
-const isHostsRoutes = (spyState: RouteSpyState): spyState is HostRouteSpyState =>
-  spyState != null && spyState.pageName === SecurityPageName.hosts;
-
-const isUsersRoutes = (spyState: RouteSpyState): spyState is UsersRouteSpyState =>
-  spyState != null && spyState.pageName === SecurityPageName.users;
+export const getBreadcrumbsForRoute = (
+  object: RouteSpyState & ObjectWithNavTabs,
+  getSecuritySolutionUrl: GetSecuritySolutionUrl,
+  isGroupedNavigationEnabled: boolean
+): ChromeBreadcrumb[] | null => {
+  const spyState: RouteSpyState = omit('navTabs', object);
 
-const isTimelinesRoutes = (spyState: RouteSpyState): spyState is TimelineRouteSpyState =>
-  spyState != null && spyState.pageName === SecurityPageName.timelines;
+  if (!spyState || !object.navTabs || !spyState.pageName) {
+    return null;
+  }
 
-const isCaseRoutes = (spyState: RouteSpyState): spyState is RouteSpyState =>
-  spyState != null && spyState.pageName === SecurityPageName.case;
+  const newMenuLeadingBreadcrumbs = getLeadingBreadcrumbsForSecurityPage(
+    spyState.pageName as SecurityPageName,
+    getSecuritySolutionUrl,
+    object.navTabs,
+    isGroupedNavigationEnabled
+  );
 
-const isAdminRoutes = (spyState: RouteSpyState): spyState is AdministrationRouteSpyState =>
-  spyState != null && spyState.pageName === SecurityPageName.administration;
+  // last newMenuLeadingBreadcrumbs is the current page
+  const pageBreadcrumb = newMenuLeadingBreadcrumbs[newMenuLeadingBreadcrumbs.length - 1];
+  const siemRootBreadcrumb = newMenuLeadingBreadcrumbs[0];
 
-const isRulesRoutes = (spyState: RouteSpyState): spyState is AdministrationRouteSpyState =>
-  spyState != null && spyState.pageName === SecurityPageName.rules;
+  const leadingBreadcrumbs = isGroupedNavigationEnabled
+    ? newMenuLeadingBreadcrumbs
+    : [siemRootBreadcrumb, pageBreadcrumb];
 
-// eslint-disable-next-line complexity
-export const getBreadcrumbsForRoute = (
-  object: RouteSpyState & TabNavigationProps,
-  getUrlForApp: GetUrlForApp
-): ChromeBreadcrumb[] | null => {
-  const spyState: RouteSpyState = omit('navTabs', object);
-  const landingPath = getUrlForApp(APP_UI_ID, { deepLinkId: SecurityPageName.landing });
+  // Cases plugin controls breadcrumbs
+  if (isCaseRoutes(spyState)) {
+    return null;
+  }
 
-  const siemRootBreadcrumb: ChromeBreadcrumb = {
-    text: APP_NAME,
-    href: getAppLandingUrl(landingPath),
-  };
-  if (isHostsRoutes(spyState) && object.navTabs) {
-    const tempNav: SearchNavTab = { urlKey: 'host', isDetailPage: false };
-    let urlStateKeys = [getOr(tempNav, spyState.pageName, object.navTabs)];
-    if (spyState.tabName != null) {
-      urlStateKeys = [...urlStateKeys, getOr(tempNav, spyState.tabName, object.navTabs)];
+  // Admin URL works differently. All admin pages are under '/administration'
+  if (isAdminRoutes(spyState)) {
+    if (isGroupedNavigationEnabled) {
```

Comment:
Not sure if it's worth it here, but `concat` tends to be a bit more performant than the double spread: `leadingBreadcrumbs.concat(getAdminBreadcrumbs(spyState))`. Don't think it matters at this scale though

Reply:
Hey Michael, the concatenated array has a max length of 4. So I am keeping the current implementation because there is no performance concern for this scenario.

Thanks for the valuable tip!

Reply:
Agreed, and thank you!